### ISSUE TITLE
How to access session object from rspec tests?

### ISSUE THREAD

**Issue by zhulik** [telegram-bot-rb/telegram-bot#58]

**Comment by zhulik** [comment]
I did't find any ways to do it in a 'right' way, so I have mocked the session method in controller.

```
RSpec.shared_context 'telegram/bot/session' do
  def override_session(session)
    allow_any_instance_of(described_class).to receive(:session).and_return(session)
  end
end
```

**Comment by printercu** [comment]
Rails integration tests delegates session to `@request` (in `ActionDispatch::TestProcess`). Same way it's possible to have something like this:

```ruby
  def session
    described_class.new(bot, from: from.stringify_keys, chat: chat.stringify_keys).
      send(:session)
  end
```

This way it'll test `session_key` as well. I'm not sure if this should be part of gem, as session key may depend on other params. May be just add a note to wiki. And rails encourage testing external effects (responses) in request specs, instead of internal entities.

**Comment by zhulik** [comment]
Due to rails documentation, @request is not available before request is made.  Yes, testing session content in requests is not a good idea for web, but I think it's very important for bots.

**Comment by printercu** [comment]
Yes, so it's usually required to init session with `get '/'` to use `expect { ... }.to change { session[:smth] }...`.

Does my snippet help?

**Comment by printercu** [comment]
@zhulik any updates on this? Is the issue resolved?

**Comment by zhulik** [comment]
I suppose it will crash, because session method is delegated to `@request` which does not exist before `get`.

**Comment by printercu** [comment]
Bot's session does not use request, more on this in readme. Please check my first comment: https://github.com/telegram-bot-rb/telegram-bot/issues/58#issuecomment-354984895. Provided solution does not rely on `@request`. If it does not work for you, I'm happy to receive other proposals.

**Comment by printercu** [comment]
any updates on this? Is the issue resolved?

**Comment by printercu** [comment]
Closing. Feel free to ping me to reopen.
